### ISSUE TITLE
BehaviorSubject + switchMap does not emit when expected after 24.1 upgrade

### ISSUE THREAD

**Issue by Mike278** [ReactiveX/rxdart#477]
Tried again to upgrade rxdart from 23.1 to 24.1, but some stream-powered widgets in our app stopped loading.

I've reduced it down to this case:

```dart
test("rxdart 23.1 -> 24.1", () async {
  final a = BehaviorSubject.seeded('a');
  final bug = a.switchMap((_) => BehaviorSubject.seeded('b'));
  bug.listen((b) => print('data! $b'));
  print(await bug.first);
  print(await bug.first); // hangs here with rxdart 24.1, passes with rxdart 23.1
});
```

**Comment by hoc081098** [comment]
Hi @Mike278, IMO
- Because inside `BehaviorSubject`, we use `StartWithStreamTransformer` to replay latest value.
- - RxDart 23.1: `StartWithStreamTransformer` adds value synchronously
- - RxDart 24.1: `StartWithStreamTransformer` adds value based on `sync` param in `BehaviorSubject.seeded(value, { bool sync: false }). By default, `sync` is false, so event added asynchronously.
- Fix: make `BehaviorSubject` is `sync` eg: `BehaviorSubject.seeded(seedValue, sync: true)`

**Comment by Mike278** [comment]
Thanks for the reply!

Adding `sync: true` just make it hang one line earlier at the first `​print​(​await​ bug.first);` - did something different happen when you ran it?

I'm not sure I understand how StreamController's `sync` parameter would have helped here - the problem is not that an event is delivered at an unexpected time, it's that the future returned by the second call to `bug.first` does not complete at all.

The bug seems to be triggered by a specific sequence of listeners. For example 24.1 has the same behavior as 23.1 if you comment out `bug.​listen​((b) ​=>​ ​print​(​'data! $​b​'​));` or swap it with the line below:

```dart
BehaviorSubject<String> a;
Stream<String> bug;
setUp(() {
  a = BehaviorSubject.seeded('a');
  bug = a.switchMap((_) => BehaviorSubject.seeded('b'));
});

test("v1", () async {
  bug.listen((b) => print('data! $b'));
  print(await bug.first);
  print(await bug.first); // hangs here with rxdart 24.1, passes with rxdart 23.1
});

test("v2", () async {
  // bug.listen((b) => print('data! $b'));
  print(await bug.first);
  print(await bug.first); // works
});

test("v3", () async {
  print(await bug.first);
  bug.listen((b) => print('data! $b'));
  print(await bug.first); // works
});
```

I also discovered a possibly related bug in #478

**Comment by benoitskipr** [comment]
thanks for the example @Mike278 , I'm having this issue as well, and we shouldn't need to specify any arguments, cause it is default behavior (get it?) that all subsequent subscriptions get the last stored value, right ? 
whether we've called `listen()` or not, `await first` shouldn`t  be too different from a listen() anyway

**Comment by benoitskipr** [comment]
having the same issue, does not feel like normal behaviorsubject ... behavior

**Comment by Mike278** [comment]
Any movement on this bug? It prevents us from upgrading rxdart.

**Comment by frankpepermans** [comment]
I'm looking into this, but it's not an easy fix, plus the current covid crisis & regular workload don't make it easy to commit a lot of time to rxdart at the moment :/

**Comment by Mike278** [comment]
Thanks @frankpepermans, appreciate the update :)

**Comment by frankpepermans** [comment]
pre-release is up: https://pub.dev/packages/rxdart/versions/0.25.0-beta

**Comment by Mike278** [comment]
I gave it a try and while the repro test now passes, I'm still seeing the issue with actual widgets in our app. I'll try to get a minimal repro tomorrow.

**Comment by Mike278** [comment]
Finally had a chance to look at this again and was able to reduce it to this case: 

```dart
test('rxdart ^0.25.0-beta', () async {
  final bug = BehaviorSubject.seeded('seeded')
    .map((_) => 'map')
    .switchMap((_) => BehaviorSubject.seeded('switchMap'))
    ..listen(print);
  await pumpEventQueue();
  expect(await bug.first, 'switchMap'); // future never completes
});
```

with:
```yaml
# pubspec.yaml:
dependency_overrides:
  rxdart:
    git:
      url: https://github.com/ReactiveX/rxdart.git
      ref: master

# pubspec.lock
rxdart:
  dependency: "direct main"
  description:
    path: "."
    ref: master
    resolved-ref: "4dcefadf6f72cc5061f7b9346e93179cfd2515d6"
    url: "https://github.com/ReactiveX/rxdart.git"
  source: git
  version: "0.25.0-beta"
```

**Comment by hoc081098** [comment]
hey @Mike278 
I think because `map` methods make the `BehaviorSubject` becomes a Broadcast Stream, so we lose `replay latest value` behavior
```dart
Stream<S> map<S>(S convert(T event)) {
    return new _MapStream<T, S>(this, convert);
}
```
Temporary workaround: call `shareValue` after `map`
```dart
BehaviorSubject.seeded('seeded')
    .map((_) => 'map') 
    .shareValue()
    .switchMap((_) => BehaviorSubject.seeded('switchMap'))
      ..listen(print);
```
